[Sky] 미션3 : 체스판  초기화

aksmf1442 opened the pull request:
안녕하세요! Sky입니다.
이번에도 리뷰 잘 부탁드립니다!

### 미션3 변경사항
* 먼저 가장 고민되었던 점이 'Pawn의 representation을 어떻게 하면 좋을까?' 라는 생각이었는데 처음에는 Color의 enum을 White('p'), Black('P') 이런식으로 구현해볼까? 라는 생각을 했지만 확장성을 고려했을 때 이 방법은 좋지 않다는 생각이 들어서 따로 Representation enum을 만들어서 Representation {P('P'), p('p')}과 같이 구현했습니다. 그런데 이 방법도 그렇게 좋다는 생각이 들지 않아서 계속 고민하다가 일단은 넘어갔습니다.

* 위에 적은 것처럼 변경하니까 Pawn을 관리하는 List인 pawns는 따로 필요없을 것 같아서 제거하고, whitePawns와 blackPawns List를 따로 추가해주고, 이에 따라서 add(), size(), findPawn() 메서드들을 white, black별로 역할을 분리해서 메서드를 새로 구현했습니다.

* Board의 print()를 구현하기 위해서 처음에는 if문을 사용했었는데 이렇게 구현하니까 오히려 보기 안 좋아서 하드코딩으로 변경했습니다.

* ChessGame 클래스를 추가하고 main 메서드에 while문을 이용해서 게임을 하는 부분을 구현했었는데 main 메서드가 길어지는게 보기 안 좋아서  실제 게임을 하는 부분을 gameStart() 메서드를 추가하고 이 곳으로 분리했습니다.

aksmf1442 committed:
Add Representation Enum to Pawn, Add representation test to PawnTest

폰의 색깔에 따른 표현방식을 추가하기 위해서 Pawn Class에 Representation Enum을 사용해서 생성자로 표현방식을
받을 수 있도록 하고, 이를 생성자로 받으면서 생기는 사이드 이펙트들을 처리해줬다.
Pawn Class 안에 Presentation Enum의 코드량이 많아져서 밖으로 따로 enum을
뺄까도 생각했는데 일단은 Pawn에서 표현하는 방식이니까 따로 빼지는 않았다.

aksmf1442 committed:
Add initialize method to Board

체스판에 white, black 폰들을 초기화하기 위해서 initailize 메서드를 Board 클래스에 추가

aksmf1442 committed:
Refactor Board and BoardTest

Board 클래스의 중복을 제거하고, 메서드 분리를 위해서 add, size, find 메서드들을 white, black별로 분리
그에 따라서 BoardTest의 내용도 수정하고 네이밍도 바꿈

aksmf1442 committed:
Add print method to Board

콘솔 화면에 출력하기 위해서 Board 클래스에 print 메서드 추가

aksmf1442 committed:
Add ChessGame for start

게임의 진행을 위해서 ChessGame 클래스에 main 메서드를 만들고 start를 입력하면 게임 시작 end를 입력하면 게임을 끝나게 한다.

aksmf1442 committed:
Simply ChessGame, Board

ChessGame 클래스의 main 메서드를 깔끔하게 표현하고 싶어서 게임이 동작하는 부분은 따로 gameStart라는 메서드로 분리했다.
Board 클래스의 print 부분을 if문을 사용해서 구현했었는데 오히려 하드코딩하는 것보다 더 지저분한 느낌이 들어서 하드코딩으로 변경

aksmf1442 committed:
Simply gameStart() to ChessGame

ChessGame 클래스에서 else if를 제거하고 if문으로 변경

honux77 commented on src/main/java/net/sky/chess/Board.java:
```diff
@@ -4,25 +4,76 @@
 
-    private List<Pawn> pawns;
+    private List<Pawn> whitePawns;
+
```
공백은 없어도 될 듯?

aksmf1442 replied:
넵. 공백 제거했습니다!

honux77 commented on src/main/java/net/sky/chess/Board.java:
```diff
@@ -4,25 +4,76 @@
 
     public Board() {
-        pawns = new ArrayList<>();
+        whitePawns = new ArrayList<>();
```
이정도라면 생성자보다 직접 초기화를 해 주는 것도 괜찮습니다.

aksmf1442 replied:
아 그렇군요. 무조건 생성자에서 초기화 시켜주면 좋은건줄 알았어요! 조금 더 공부해봐야겠네요.

honux77 commented on src/main/java/net/sky/chess/Board.java:
```diff
@@ -4,25 +4,76 @@
-        pawns.add(pawn);
+    public void initialize() {
+        for (int i = 0; i < 8; i++) {
+            addWhitePawn(new Pawn(Color.WHITE, Representation.p));
```
지금 단계에서 굳이 Representaion enum은 필요하지 않은 것 같아요. 단순하게 캐릭터 상수로 표현했으면 어땠을까요?
둘 중에 어떤게 가독성이 좋아지는지 어떤게 코드가 간결해지는지 한번 비교해 보세요.

honux77 replied:
지금 코드라면 WHITE이면서 P인 폰을 생성할 수 있게 되는데, 그것도 괜찮을까요?

aksmf1442 replied:
처음에 Color enum에 바로 representation을 표현 해주는건 어떨까 싶었는데, 그러면 Color enum에 종속되는 기분이 들어서 따로 Representation enum을 만드는 방식으로 했더니 오히려 가독성도 안 좋았던 것 같네요.

aksmf1442 replied:
일단 지금 단계에서 깔끔해보이는건 제가 생각하기엔 Color enum에 representation 변수를 만들어주는게 더 가독성이 좋을 것 같아서 일단 그렇게 변경했습니다!

honux77 commented on src/main/java/net/sky/chess/Board.java:
```diff
@@ -4,25 +4,76 @@
-        return pawns.get(idx);
+    public String getBlackPawnsResult() {
+        StringBuilder result = new StringBuilder();
+        blackPawns.forEach((x) -> result.append(x.getRepresentation().getValue()));
```
스트림 forEach는 꼭 필요한 경우에만 쓰는게 좋아요. 지금은 단순 개선된 for loop로 충분해 보입니다. 관련해서 공부해 보세요.

honux77 replied:
x.getRepresentation().getValue() 보다 x.getRepresentation()이 한번에 "P"를 리턴해도 좋을 것 같습니다.

aksmf1442 replied:
forEach를 사용하면 가독성이 좋을 것 같아서 사용했는데, 굳이 쓸 필요가 없었던 것 같네요. 개선된 for loop로 변경했습니다.
스트림에 대해서는 '모던 자바 인 액션'이라는 책을 지금 보고 있는데 조금 더 깊게 공부할 필요가 있는 것 같네요.

honux77 reviewed (changes requested):
잘 하셨어요. 다만 제 느낌은 클래스들간의 결합도가 살짝 높은 느낌(?)입니다만..
외부 API가 내부 구조에 영향을 덜 받게 개선을 하면 좋겠다는 생각이 들어요. 

조금 더 깔끔하게 개선을 해 보는 것도 좋을 것 같습니다.

aksmf1442 committed:
Refactor Pawn and Board, Add public enum Color

Pawn 클래스 내부에 있던 Representation enum을 제거하고 Color enum의 representation 변수를 만들어줘서 White('p')와 같은 형식으로
변수를 관리할 수 있게 변경했다.
클래스 간의 결합도를 줄이기 위해서 Pawn 클래스 내부에 있던 Color enum을 분리해서 public enum으로 변경했다.
Board 클래스의 getWhite(Black)PawnsResult() 메서드에서 forEach문을 사용하고 있었느데 이를 개선된 for loop문으로 변경했다.

aksmf1442 commented:
리뷰 감사합니다!

honux77 reviewed (approved):
👏
